roachtest: schemachange/index/tpcc/w=1000 failed

[(roachtest).schemachange/index/tpcc/w=1000 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2578434&tab=buildLog) on [master@03797b17417ae34451537f8f76d66ac69dba2d07](https://github.com/cockroachdb/cockroach/commits/03797b17417ae34451537f8f76d66ac69dba2d07):

```
		  -- stack trace:
		  | main.(*cluster).RunE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2292
		  | main.(*cluster).Run
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2212
		  | main.setupTPCC.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:118
		  | main.setupTPCC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:134
		  | main.runTPCC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:145
		  | main.makeIndexAddTpccTest.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/schemachange.go:302
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:760
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (2) output in run_120805.207_n1_cockroach_workload_fixtures_import_tpcc
		Wraps: (3) /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-2578434-1610521418-78-n5cpu16:1 -- ./cockroach workload fixtures import tpcc --warehouses=1000  returned
		  | stderr:
		  | I210113 12:08:05.970932 1 ccl/workloadccl/fixture.go:342  starting import of 9 tables
		  | I210113 12:08:08.372570 89 ccl/workloadccl/fixture.go:472  imported 53 KiB in warehouse table (1000 rows, 0 index entries, took 2.390063415s, 0.02 MiB/s)
		  | I210113 12:08:12.897213 95 ccl/workloadccl/fixture.go:472  imported 7.8 MiB in item table (100000 rows, 0 index entries, took 6.914379474s, 1.12 MiB/s)
		  | I210113 12:08:15.097980 90 ccl/workloadccl/fixture.go:472  imported 1006 KiB in district table (10000 rows, 0 index entries, took 9.115584831s, 0.11 MiB/s)
		  | I210113 12:08:19.442763 94 ccl/workloadccl/fixture.go:472  imported 126 MiB in new_order table (9000000 rows, 0 index entries, took 13.459993787s, 9.33 MiB/s)
		  | I210113 12:09:06.465858 92 ccl/workloadccl/fixture.go:472  imported 2.2 GiB in history table (30000000 rows, 0 index entries, took 1m0.483100696s, 36.47 MiB/s)
		  | I210113 12:09:12.647954 93 ccl/workloadccl/fixture.go:472  imported 1.6 GiB in order table (30000000 rows, 30000000 index entries, took 1m6.665487556s, 23.93 MiB/s)
		  | I210113 12:10:50.859260 91 ccl/workloadccl/fixture.go:472  imported 17 GiB in customer table (30000000 rows, 30000000 index entries, took 2m44.876846222s, 107.09 MiB/s)
		  | I210113 12:12:37.936432 96 ccl/workloadccl/fixture.go:472  imported 30 GiB in stock table (100000000 rows, 0 index entries, took 4m31.953740826s, 113.05 MiB/s)
		  | I210113 12:12:41.032927 114 ccl/workloadccl/fixture.go:472  imported 17 GiB in order_line table (300011520 rows, 0 index entries, took 4m35.049538069s, 62.60 MiB/s)
		  | I210113 12:12:41.069432 1 ccl/workloadccl/fixture.go:351  imported 68 GiB bytes in 9 tables (took 4m35.098330057s, 252.84 MiB/s)
		  | I210113 12:12:42.042750 1 ccl/workloadccl/cliccl/fixtures.go:355  fixture is restored; now running consistency checks (ctrl-c to abort)
		  | I210113 12:12:42.087079 1 workload/tpcc/tpcc.go:389  check 3.3.2.1 took 44.277056ms
		  | I210113 12:13:04.075991 1 workload/tpcc/tpcc.go:389  check 3.3.2.2 took 21.988833635s
		  | I210113 12:13:06.115460 1 workload/tpcc/tpcc.go:389  check 3.3.2.3 took 2.039416561s
		  | I210113 12:16:11.060043 1 workload/tpcc/tpcc.go:389  check 3.3.2.4 took 3m4.944520501s
		  | I210113 12:17:01.056623 1 workload/tpcc/tpcc.go:389  check 3.3.2.5 took 49.996507159s
		  | Error: check failed: 3.3.2.5: right EXCEPT left returned nonzero results.
		  | Error: COMMAND_PROBLEM: exit status 1
		  | (1) COMMAND_PROBLEM
		  | Wraps: (2) Node 1. Command with error:
		  |   | ```
		  |   | ./cockroach workload fixtures import tpcc --warehouses=1000
		  |   | ```
		  | Wraps: (3) exit status 1
		  | Error types: (1) errors.Cmd (2) *hintdetail.withDetail (3) *exec.ExitError
		  |
		  | stdout:
		Wraps: (4) exit status 20
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *main.withCommandDetails (4) *exec.ExitError

```

<details><summary>More</summary><p>


Artifacts: [/schemachange/index/tpcc/w=1000](https://teamcity.cockroachdb.com/viewLog.html?buildId=2578434&tab=artifacts#/schemachange/index/tpcc/w=1000)
Related:
- #58837 roachtest: schemachange/index/tpcc/w=1000 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #54304 roachtest: schemachange/index/tpcc/w=1000 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.2)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aschemachange%2Findex%2Ftpcc%2Fw%3D1000.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
		  | E210117 12:30:34.988061 1 workload/cli/run.go:472  error in stockLevel: ERROR: rpc error: code = Canceled desc = context canceled (SQLSTATE XXUUU)
		  | E210117 12:30:36.722013 1 workload/cli/run.go:472  error in newOrder: ERROR: rpc error: code = Unavailable desc = transport is closing (SQLSTATE XXUUU)
		  | E210117 12:30:49.537707 1 workload/cli/run.go:472  error in delivery: ERROR: result is ambiguous (error=rpc error: code = Unavailable desc = transport is closing [exhausted]) (SQLSTATE 40003)
```

Context cancelled errors in the workload?

--- a/pkg/sql/backfill.go
+++ b/pkg/sql/backfill.go
@@ -23,6 +23,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/sql/backfill"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkv"
@@ -70,22 +71,19 @@ const (
 	// over many ranges.
 	indexTxnBackfillChunkSize = 100
 
-	// indexBackfillBatchSize is the maximum number of index entries we attempt to
-	// fill in a single index batch before queueing it up for ingestion and
-	// progress reporting in the index backfiller processor.
-	//
-	// TODO(adityamaru): This should live with the index backfiller processor
-	// logic once the column backfiller is reworked. The only reason this variable
-	// is initialized here is to maintain a single testing knob
-	// `BackfillChunkSize` to control both the index and column backfill chunking
-	// behavior, and minimize test complexity. Should this be a cluster setting? I
-	// would hope we can do a dynamic memory based adjustment of this number in
-	// the processor.
-	indexBackfillBatchSize = 5000
-
 	// checkpointInterval is the interval after which a checkpoint of the
 	// schema change is posted.
 	checkpointInterval = 2 * time.Minute
+)
+
+// indexBackfillBatchSize is the maximum number of rows we construct index
+// entries for before we attempt to fill in a single index batch before queueing
+// it up for ingestion and progress reporting in the index backfiller processor.
+var indexBackfillBatchSize = settings.RegisterIntSetting(
+	"bulkio.index_backfill.batch_size",
+	"the number of rows for which we construct index entries in a single batch",
+	50000,
+	settings.NonNegativeInt, /* validateFn */
 )
 
 var _ sort.Interface = columnsByID{}
@@ -949,7 +947,6 @@ func (sc *SchemaChanger) distIndexBackfill(
 	targetSpans []roachpb.Span,
 	addedIndexes []descpb.IndexID,
 	filter backfill.MutationFilter,
-	indexBackfillBatchSize int64,
 ) error {
 	readAsOf := sc.clock.Now()
 
@@ -1025,7 +1022,8 @@ func (sc *SchemaChanger) distIndexBackfill(
 		evalCtx = createSchemaChangeEvalCtx(ctx, sc.execCfg, txn.ReadTimestamp(), sc.ieFactory)
 		planCtx = sc.distSQLPlanner.NewPlanningCtx(ctx, &evalCtx, nil /* planner */, txn,
 			true /* distribute */)
-		chunkSize := sc.getChunkSize(indexBackfillBatchSize)
+		indexBatchSize := indexBackfillBatchSize.Get(&sc.execCfg.Settings.SV)
+		chunkSize := sc.getChunkSize(indexBatchSize)
 		spec, err := initIndexBackfillerSpec(*tableDesc.TableDesc(), readAsOf, chunkSize, addedIndexes)
 		if err != nil {
 			return err
@@ -1778,7 +1776,7 @@ func (sc *SchemaChanger) backfillIndexes(
 	}
 
 	if err := sc.distIndexBackfill(
-		ctx, version, addingSpans, addedIndexes, backfill.IndexMutationFilter, indexBackfillBatchSize,
+		ctx, version, addingSpans, addedIndexes, backfill.IndexMutationFilter,
 	); err != nil {
 		return err
 	}

--- a/pkg/sql/backfill/backfill.go
+++ b/pkg/sql/backfill/backfill.go
@@ -52,17 +52,9 @@ func IndexMutationFilter(m descpb.DescriptorMutation) bool {
 	return m.GetIndex() != nil && m.Direction == descpb.DescriptorMutation_ADD
 }
 
-// backfiller is common to a ColumnBackfiller or an IndexBackfiller.
-type backfiller struct {
-	fetcher row.Fetcher
-	alloc   rowenc.DatumAlloc
-}
-
 // ColumnBackfiller is capable of running a column backfill for all
 // updateCols.
 type ColumnBackfiller struct {
-	backfiller
-
 	added   []descpb.ColumnDescriptor
 	dropped []descpb.ColumnDescriptor
 
@@ -70,6 +62,9 @@ type ColumnBackfiller struct {
 	updateCols  []descpb.ColumnDescriptor
 	updateExprs []tree.TypedExpr
 	evalCtx     *tree.EvalContext
+
+	fetcher row.Fetcher
+	alloc   rowenc.DatumAlloc
 
 	// mon is a memory monitor linked with the ColumnBackfiller on creation.
 	mon *mon.BytesMonitor
@@ -409,8 +404,6 @@ type muBoundAccount struct {
 
 // IndexBackfiller is capable of backfilling all the added index.
 type IndexBackfiller struct {
-	backfiller
-
 	added []*descpb.IndexDescriptor
 	// colIdxMap maps ColumnIDs to indices into desc.Columns and desc.Mutations.
 	colIdxMap catalog.TableColMap
@@ -444,6 +437,10 @@ type IndexBackfiller struct {
 	// backfilled.
 	indexesToEncode []*descpb.IndexDescriptor
 
+	valNeededForCol util.FastIntSet
+
+	alloc rowenc.DatumAlloc
+
 	// mon is a memory monitor linked with the IndexBackfiller on creation.
 	mon            *mon.BytesMonitor
 	muBoundAccount muBoundAccount
@@ -473,7 +470,7 @@ func (ib *IndexBackfiller) InitForLocalUse(
 	ib.initCols(desc)
 
 	// Initialize ib.added.
-	valNeededForCol := ib.initIndexes(desc)
+	ib.valNeededForCol = ib.initIndexes(desc)
 
 	predicates, colExprs, referencedColumns, err := constructExprs(
 		ctx, desc, ib.added, ib.cols, ib.addedCols, ib.computedCols, evalCtx, semaCtx,
@@ -485,10 +482,10 @@ func (ib *IndexBackfiller) InitForLocalUse(
 	// Add the columns referenced in the predicate to valNeededForCol so that
 	// columns necessary to evaluate the predicate expression are fetched.
 	referencedColumns.ForEach(func(col descpb.ColumnID) {
-		valNeededForCol.Add(ib.colIdxMap.GetDefault(col))
+		ib.valNeededForCol.Add(ib.colIdxMap.GetDefault(col))
 	})
 
-	return ib.init(evalCtx, predicates, colExprs, valNeededForCol, desc, mon)
+	return ib.init(evalCtx, predicates, colExprs, mon)
 }
 
 // constructExprs is a helper to construct the index and column expressions
@@ -594,7 +591,7 @@ func (ib *IndexBackfiller) InitForDistributedUse(
 	ib.initCols(desc)
 
 	// Initialize ib.added.
-	valNeededForCol := ib.initIndexes(desc)
+	ib.valNeededForCol = ib.initIndexes(desc)
 
 	evalCtx := flowCtx.NewEvalCtx()
 	var predicates map[descpb.IndexID]tree.TypedExpr
@@ -630,15 +627,14 @@ func (ib *IndexBackfiller) InitForDistributedUse(
 	// Add the columns referenced in the predicate to valNeededForCol so that
 	// columns necessary to evaluate the predicate expression are fetched.
 	referencedColumns.ForEach(func(col descpb.ColumnID) {
-		valNeededForCol.Add(ib.colIdxMap.GetDefault(col))
+		ib.valNeededForCol.Add(ib.colIdxMap.GetDefault(col))
 	})
 
-	return ib.init(evalCtx, predicates, colExprs, valNeededForCol, desc, mon)
+	return ib.init(evalCtx, predicates, colExprs, mon)
 }
 
 // Close releases the resources used by the IndexBackfiller.
 func (ib *IndexBackfiller) Close(ctx context.Context) {
-	ib.fetcher.Close(ctx)
 	if ib.mon != nil {
 		ib.muBoundAccount.Lock()
 		ib.muBoundAccount.boundAccount.Close(ctx)
@@ -728,8 +724,6 @@ func (ib *IndexBackfiller) init(
 	evalCtx *tree.EvalContext,
 	predicateExprs map[descpb.IndexID]tree.TypedExpr,
 	colExprs map[descpb.ColumnID]tree.TypedExpr,
-	valNeededForCol util.FastIntSet,
-	desc catalog.TableDescriptor,
 	mon *mon.BytesMonitor,
 ) error {
 	ib.evalCtx = evalCtx
@@ -750,32 +744,13 @@ func (ib *IndexBackfiller) init(
 		ib.types[i] = ib.cols[i].Type
 	}
 
-	tableArgs := row.FetcherTableArgs{
-		Desc:            desc,
-		Index:           desc.GetPrimaryIndex().IndexDesc(),
-		ColIdxMap:       ib.colIdxMap,
-		Cols:            ib.cols,
-		ValNeededForCol: valNeededForCol,
-	}
-
 	// Create a bound account associated with the index backfiller monitor.
 	if mon == nil {
 		return errors.AssertionFailedf("no memory monitor linked to IndexBackfiller during init")
 	}
 	ib.mon = mon
 	ib.muBoundAccount.boundAccount = mon.MakeBoundAccount()
-
-	return ib.fetcher.Init(
-		evalCtx.Context,
-		evalCtx.Codec,
-		false, /* reverse */
-		descpb.ScanLockingStrength_FOR_NONE,
-		descpb.ScanLockingWaitPolicy_BLOCK,
-		false, /* isCheck */
-		&ib.alloc,
-		mon,
-		tableArgs,
-	)
+	return nil
 }
 
 // BuildIndexEntriesChunk reads a chunk of rows from a table using the span sp
@@ -816,7 +791,29 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 	// during the scan. Index entries in the new index are being
 	// populated and deleted by the OLTP commands but not otherwise
 	// read or used
-	if err := ib.fetcher.StartScan(
+	tableArgs := row.FetcherTableArgs{
+		Desc:            tableDesc,
+		Index:           tableDesc.GetPrimaryIndex().IndexDesc(),
+		ColIdxMap:       ib.colIdxMap,
+		Cols:            ib.cols,
+		ValNeededForCol: ib.valNeededForCol,
+	}
+	var fetcher row.Fetcher
+	if err := fetcher.Init(
+		ib.evalCtx.Context,
+		ib.evalCtx.Codec,
+		false, /* reverse */
+		descpb.ScanLockingStrength_FOR_NONE,
+		descpb.ScanLockingWaitPolicy_BLOCK,
+		false, /* isCheck */
+		&ib.alloc,
+		ib.mon,
+		tableArgs,
+	); err != nil {
+		return nil, nil, 0, err
+	}
+	defer fetcher.Close(ctx)
+	if err := fetcher.StartScan(
 		ctx, txn, []roachpb.Span{sp}, true /* limitBatches */, initBufferSize,
 		traceKV, false, /* forceProductionKVBatchSize */
 	); err != nil {
@@ -860,7 +857,7 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 		return nil
 	}
 	for i := int64(0); i < chunkSize; i++ {
-		encRow, _, _, err := ib.fetcher.NextRow(ctx)
+		encRow, _, _, err := fetcher.NextRow(ctx)
 		if err != nil {
 			return nil, nil, 0, err
 		}
@@ -962,7 +959,12 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 	ib.ShrinkBoundAccount(ctx, shrinkSize)
 	memUsedPerChunk -= shrinkSize
 
-	return entries, ib.fetcher.Key(), memUsedPerChunk, nil
+	var resumeKey roachpb.Key
+	if fetcher.Key() != nil {
+		resumeKey = make(roachpb.Key, len(fetcher.Key()))
+		copy(resumeKey, fetcher.Key())
+	}
+	return entries, resumeKey, memUsedPerChunk, nil
 }
 
 // RunIndexBackfillChunk runs an index backfill over a chunk of the table

--- a/pkg/sql/index_backfiller.go
+++ b/pkg/sql/index_backfiller.go
@@ -130,8 +130,8 @@ func (ib *IndexBackfillPlanner) plan(
 			true /* distribute */)
 		// TODO(ajwerner): Adopt util.ConstantWithMetamorphicTestRange for the
 		// batch size. Also plumb in a testing knob.
-		spec, err := initIndexBackfillerSpec(
-			*td.TableDesc(), readAsOf, indexBackfillBatchSize, indexesToBackfill)
+		chunkSize := indexBackfillBatchSize.Get(&ib.execCfg.Settings.SV)
+		spec, err := initIndexBackfillerSpec(*td.TableDesc(), readAsOf, chunkSize, indexesToBackfill)
 		if err != nil {
 			return err
 		}

--- a/pkg/sql/indexbackfiller_test.go
+++ b/pkg/sql/indexbackfiller_test.go
@@ -522,9 +522,8 @@ INSERT INTO foo VALUES (1), (10), (100);
 			tableID, 1, execCfg.NodeID.SQLInstanceID(), s0.DB(), lm, jr, &execCfg, settings)
 		changer.SetJob(j)
 		spans := []roachpb.Span{table.IndexSpan(keys.SystemSQLCodec, test.indexToBackfill)}
-		require.NoError(t, changer.TestingDistIndexBackfill(
-			ctx, table.GetVersion(), spans, []descpb.IndexID{test.indexToBackfill}, backfill.IndexMutationFilter, 10,
-		))
+		require.NoError(t, changer.TestingDistIndexBackfill(ctx, table.GetVersion(), spans,
+			[]descpb.IndexID{test.indexToBackfill}, backfill.IndexMutationFilter))
 
 		// Make the mutation complete, then read the index and validate that it
 		// has the expected contents.

--- a/pkg/sql/schema_changer_helpers_test.go
+++ b/pkg/sql/schema_changer_helpers_test.go
@@ -27,10 +27,8 @@ func (sc *SchemaChanger) TestingDistIndexBackfill(
 	targetSpans []roachpb.Span,
 	addedIndexes []descpb.IndexID,
 	filter backfill.MutationFilter,
-	indexBackfillBatchSize int64,
 ) error {
-	return sc.distIndexBackfill(
-		ctx, version, targetSpans, addedIndexes, filter, indexBackfillBatchSize)
+	return sc.distIndexBackfill(ctx, version, targetSpans, addedIndexes, filter)
 }
 
 // SetJob sets the job.

